staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,189 @@
+	constants "k8s.io/apimachinery/pkg/util/portforward"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/klog/v2"
+)
+
+// TunnelingHandler is a handler which tunnels SPDY through WebSockets.
+type TunnelingHandler struct {
+	// Used to communicate between upstream SPDY and downstream tunnel.
+	upgradeHandler http.Handler
+}
+
+// NewTunnelingHandler is used to create the tunnel between an upstream
+// SPDY connection and a downstream tunneling connection through the stored
+// UpgradeAwareProxy.
+func NewTunnelingHandler(upgradeHandler http.Handler) *TunnelingHandler {
+	return &TunnelingHandler{upgradeHandler: upgradeHandler}
+}
+
+// ServeHTTP uses the upgradeHandler to tunnel between a downstream tunneling
+// connection and an upstream SPDY connection. The tunneling connection is
+// a wrapped WebSockets connection which communicates SPDY framed data.
+func (h *TunnelingHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	klog.V(4).Infoln("TunnelingHandler ServeHTTP")
+	// First, terminate the websocket connection, creating the net.Conn
+	// that will be used to tunnel SPDY.
+	var upgrader = gwebsocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
+			return true // Accepting all requests
+		},
+		Subprotocols: []string{
+			constants.TunnelPortForwardV1,
+		},
+	}
+	conn, err := upgrader.Upgrade(w, req, nil)
+	if err != nil {
+		klog.Errorf("error upgrading websocket connection: %v", err)
+		return
+	}
+	defer conn.Close() //nolint:errcheck
+	tunnelProtocol := conn.Subprotocol()
+	klog.V(4).Infof("websocket connection created: %s", tunnelProtocol)
+	tunnelingConn := portforward.NewTunnelingConnection("server", conn)
+	headerInterceptingConnection := &headerInterceptingConnection{Conn: tunnelingConn, headerBuffer: bytes.NewBuffer(nil)}
+	// Create ResponseWriter which will be hijacked to use the tunnel.
+	writer := createTunnelingResponseWriter(w, headerInterceptingConnection)
+	klog.V(4).Infoln("Tunnel spdy through websockets using the UpgradeAwareProxy")
+	protocol := strings.TrimPrefix(tunnelProtocol, constants.SpdyTunnelingPrefix)
+	h.upgradeHandler.ServeHTTP(writer, createSPDYRequest(req, protocol))
+}
+
+// createSPDYRequest modifies the passed request to remove
+// WebSockets headers and add SPDY upgrade information, including
+// the negotiated protocol (without the tunnel prefix).
+func createSPDYRequest(req *http.Request, protocol string) *http.Request {
+	clone := utilnet.CloneRequest(req)
+	// Clean up the websocket headers from the http request.
+	clone.Header.Del(wsstream.WebSocketProtocolHeader)
+	clone.Header.Del("Sec-Websocket-Key")
+	clone.Header.Del("Sec-Websocket-Version")
+	clone.Header.Del(httpstream.HeaderUpgrade)
+	// Update the http request for an upstream SPDY upgrade.
+	clone.Method = "POST"
+	clone.Body = nil // Remove the request body which is unused.
+	clone.Header.Add(httpstream.HeaderUpgrade, spdy.HeaderSpdy31)
+	clone.Header.Add(httpstream.HeaderProtocolVersion, protocol)
```

Comment:
prefer `Set` to `Add` (we don't want to combine with any existing values in these headers)

Reply:
Fixed.